_source/_posts/2022-01-13-node-deploy-to-heroku.md

```diff
@@ -0,0 +1,492 @@
+app.use(oidc.router);
+
+// view engine setup
+app.set("views", path.join(__dirname, "views"));
+app.set("view engine", "jade");
+
+app.use(logger("dev"));
+app.use(express.json());
+app.use(express.urlencoded({ extended: false }));
+app.use(cookieParser());
+app.use(express.static(path.join(__dirname, "public")));
+
+app.use(express.static(__dirname + "/node_modules/bootstrap/dist"));
+
+app.use("/", indexRouter);
+app.use("/users", usersRouter.userRoutes({ oidc: oidc }));
+
+// catch 404 and forward to error handler
+app.use(function (req, res, next) {
+  next(createError(404));
+});
+
+// error handler
+app.use(function (err, req, res, next) {
+  // set locals, only providing error in development
+  res.locals.message = err.message;
+  res.locals.error = req.app.get("env") === "development" ? err : {};
+
+  // render the error page
+  res.status(err.status || 500);
+  res.render("error");
+});
+
+module.exports = app;
+```
+
+Most of this code is boilerplate from the express-generator but the notable addition is the configuration for the Okta oidc.  First, you need to enable `dotenv` at the top of this file.  Next, you configure the `express-session` and add it to your application.  Finally, you can configure your Okta oidc middleware using the variables from `.env`.   
+
+Next, open your `routes/index.js` file and replace the code with the following.
+
+```javascript
+var express = require("express");
+var router = express.Router();
+
+/* GET home page. */
+router.get("/", function (req, res, next) {
+  res.render("index", { loggedIn: req.isAuthenticated() });
+});
+
+module.exports = router;
+```
+
+Here are you added a property for the `index.jade` template to use called `loggedIn`.  This property will help the view decide if it should display a `login` or `logout` button.  The rest of the home page will be static so there is no need for a more sophisticated model.
+
+Finally, open your `routes/users.js` file and replace the code there with the code below.
+
+```javascript
+var express = require("express");
+var router = express.Router();
+
+function userRoutes(options) {
+  const oidc = options.oidc;
+
+  router.get("/", oidc.ensureAuthenticated(), function (req, res, next) {
+    res.render("users/index", {
+      loggedIn: true,
+      title: "Express",
+      user: req.userContext.userinfo,
+    });
+  });
+
+  return router;
+}
+
+module.exports.userRoutes = userRoutes;
+```
+
+This file is where the magic from the Okta oidc middleware shines.  In your route definition, you are passing in the `oidc.ensureAuthenticated` middleware.  This function will reroute unauthenticated users to the login route you defined in your `app.js` file.  Because you didn't explicitly set this parameter it will direct the user to `/login` and trigger the authentication flow.  The model is passed to the view here also contains some information about the user that you will display to the user when they land on this page.  
```

Comment:
We should replace `oidc` mentions in the text with `OIDC`, unless of course, it is the name of a package.